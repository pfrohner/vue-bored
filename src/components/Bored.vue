<template>
  <main class="container">
    <h3 class="mb-3">Bored? There is an API for you!</h3>
    <p>Data is saved on localStorage, you can come back to check your list anytime.</p>
    <div class="position-relative" :class="{ disabled: isLoading }">
      <Spinner v-if="isLoading" class="spinner" />
      <b-tabs content-class="mt-3 tabs">
        <b-tab title="Activities" active>
          <Activity :activity="newActivity" @get="getRandomActivity" @add="addActivity" />
        </b-tab>
        <b-tab title="My list">
          <List :activities="activities" @clear="clearAllActivites" />
        </b-tab>
      </b-tabs>
    </div>
  </main>
</template>

<script>
import Activity from "@/components/Activity.vue";
import List from "@/components/List.vue";
import Spinner from "@/components/Spinner.vue";

export default {
  name: "Layout",
  data() {
    return {
      isLoading: false,
      newActivity: {},
      activities: []
    };
  },
  components: {
    Activity,
    List,
    Spinner
  },
  created() {
    this.getRandomActivity();
    this.getLocalStorageData();
  },
  methods: {
    addActivity(activity) {
      this.activities.push(activity);
      this.saveToLocalStorage(activity);
      this.showToast("Save successful!");
      this.getRandomActivity();
    },
    clearAllActivites() {
      const confirm = window.confirm("Are you sure?");
      if (confirm) {
        if (localStorage) {
          localStorage.removeItem("bored");
        }
        this.activities = [];
        this.showToast("Items deleted.");
      }
    },
    saveToLocalStorage(activity) {
      if (localStorage) {
        let items = JSON.parse(localStorage.getItem("bored"));
        if (!items) {
          items = [];
        }
        items.push(JSON.stringify(activity));
        localStorage.setItem("bored", JSON.stringify(items));
      }
    },
    getLocalStorageData() {
      if (localStorage) {
        const stored = JSON.parse(localStorage.getItem("bored"));
        if (stored) {
          this.activities = stored.map(event => JSON.parse(event));
        }
      }
    },
    showToast(message, variant = "success") {
      this.$bvToast.toast(message, {
        solid: true,
        appendToast: true,
        variant,
        toaster: "b-toaster-top-center"
      });
    },
    async getRandomActivity() {
      this.isLoading = true;
      try {
        const response = await this.$http.get("/activity");
        this.newActivity = response.data;
      } catch (e) {
        this.showToast(`Something went wrong!<br/ >${e}`, "danger");
      } finally {
        this.isLoading = false;
      }
    }
  }
};
</script>

<style scoped>
.spinner {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}
.disabled .tabs {
  opacity: 0.5;
  pointer-events: none;
}
</style>
