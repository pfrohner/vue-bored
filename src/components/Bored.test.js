import { mount, createLocalVue } from '@vue/test-utils'
import flushPromises from 'flush-promises'
import BTabs from 'bootstrap-vue' // make this global in jest config?

import Bored from './Bored.vue'

const response = {
  data: {
    activity: "Learn Express.js",
    type: "education",
    participants: 1,
  }
}

const setup = ({
  methods
} = {}) => {
  const localVue = createLocalVue()
  localVue.use(BTabs);

  return mount(Bored, {
    mocks: {
      $http: {
        get: () => Promise.resolve(response)
      }
    },
    methods,
    localVue
  })
}

describe('Bored', () => {
  it('renders correctly', () => {
    const wrapper = setup()

    expect(wrapper.element).toMatchSnapshot()
  })

  it('calls correct methods when created', () => {
    const getRandomActivity = jest.fn()
    const getLocalStorageData = jest.fn()

    setup({
      methods: {
        getRandomActivity,
        getLocalStorageData
      }
    })

    expect(getRandomActivity).toHaveBeenCalledTimes(1)
    expect(getLocalStorageData).toHaveBeenCalledTimes(1)
  })

  it('fetches a random activity from the API', async () => {
    const wrapper = setup()
    await flushPromises()

    expect(wrapper.vm.newActivity).toBe(response.data)
  })
})
