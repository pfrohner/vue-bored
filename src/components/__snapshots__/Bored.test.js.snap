// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bored renders correctly 1`] = `
<main
  class="container"
>
  <h3
    class="mb-3"
  >
    Bored? There is an API for you!
  </h3>
   
  <p>
    Data is saved on localStorage, you can come back to check your list anytime.
  </p>
   
  <div
    class="position-relative disabled"
  >
    <div
      class="sk-circle spinner"
    >
      <div
        class="sk-circle1 sk-child"
      />
       
      <div
        class="sk-circle2 sk-child"
      />
       
      <div
        class="sk-circle3 sk-child"
      />
       
      <div
        class="sk-circle4 sk-child"
      />
       
      <div
        class="sk-circle5 sk-child"
      />
       
      <div
        class="sk-circle6 sk-child"
      />
       
      <div
        class="sk-circle7 sk-child"
      />
       
      <div
        class="sk-circle8 sk-child"
      />
       
      <div
        class="sk-circle9 sk-child"
      />
       
      <div
        class="sk-circle10 sk-child"
      />
       
      <div
        class="sk-circle11 sk-child"
      />
       
      <div
        class="sk-circle12 sk-child"
      />
    </div>
     
    <div
      class="tabs"
    >
      <!---->
      <div
        class=""
      >
        <ul
          class="nav nav-tabs"
          role="tablist"
        >
          <!---->
          <!---->
        </ul>
      </div>
      <div
        class="tab-content mt-3 tabs"
      >
        <div
          aria-hidden="false"
          class="tab-pane active"
          role="tabpanel"
          tabindex="-1"
        >
          <section
            class="mb-3"
          >
            <div
              class="row text-left"
            >
              <div
                class="col-sm-6"
              >
                <p>
                  You should:
                </p>
                 
                <label
                  for="name"
                >
                  Activity
                </label>
                 
                <textarea
                  class="md-textarea form-control mb-3"
                  id="name"
                  rows="5"
                />
              </div>
               
              <div
                class="col-sm-6"
              >
                <p>
                  Activity details:
                </p>
                 
                <div
                  class="form-group"
                >
                  <label
                    for="type"
                  >
                    Type
                  </label>
                   
                  <input
                    class="form-control"
                    id="type"
                    type="text"
                  />
                </div>
                 
                <div
                  class="form-group"
                >
                  <label
                    for="participants"
                  >
                    Participants
                  </label>
                   
                  <input
                    class="form-control"
                    id="participants"
                    type="text"
                  />
                </div>
              </div>
            </div>
             
            <div
              class="row"
            >
              <div
                class="col-sm-6"
              >
                <button
                  class="btn btn-danger btn-block mb-3"
                  data-test="add-btn"
                >
                  Save for later!
                </button>
              </div>
               
              <div
                class="col-sm-6"
              >
                <button
                  class="btn btn-primary btn-block mb-3"
                  data-test="get-btn"
                >
                  Hit me with a new one!
                </button>
              </div>
            </div>
          </section>
        </div>
         
        <div
          aria-hidden="true"
          class="tab-pane"
          role="tabpanel"
          style="display: none;"
        >
          <section>
            <p
              data-test="empty-msg"
            >
              Your list is empty.
            </p>
          </section>
        </div>
        <div
          class="tab-pane active"
        />
      </div>
    </div>
  </div>
</main>
`;
